editor/store/selectors.js

```diff
@@ -1111,15 +1111,129 @@ export function getNotices( state ) {
+ * @returns {Editor.InserterItem}                   Item that appears in inserter.
+ */
+function buildInserterItemFromBlockType( state, enabledBlockTypes, blockType ) {
+	if ( ! enabledBlockTypes || ! blockType ) {
+		return null;
+	}
+
+	const blockTypeIsDisabled = Array.isArray( enabledBlockTypes ) && ! enabledBlockTypes.includes( blockType.name );
+	if ( blockTypeIsDisabled ) {
+		return null;
+	}
+
+	if ( blockType.isPrivate ) {
+		return null;
+	}
+
+	return {
+		name: blockType.name,
+		initialAttributes: {},
+		title: blockType.title,
+		icon: blockType.icon,
+		category: blockType.category,
+		keywords: blockType.keywords,
+		isDisabled: !! blockType.useOnce && getBlocks( state ).some( block => block.name === blockType.name ),
+	};
+}
+
+/**
+ * Given a reusable block, constructs an item that appears in the inserter.
+ * 
+ * @param {string[]|boolean}      enabledBlockTypes Enabled block types, or true/false to enable/disable all types.
+ * @param {Object}                reusableBlock     Reusable block, likely from getReusableBlock().
+ * @returns {Editor.InserterItem}                   Item that appears in inserter.
  */
-export function getRecentlyUsedBlocks( state ) {
-	// resolves the block names in the state to the block type settings
-	return compact( state.preferences.recentlyUsedBlocks.map( blockType => getBlockType( blockType ) ) );
+function buildInserterItemFromReusableBlock( enabledBlockTypes, reusableBlock ) {
+	if ( ! enabledBlockTypes || ! reusableBlock ) {
+		return null;
+	}
+
+	const blockTypeIsDisabled = Array.isArray( enabledBlockTypes ) && ! enabledBlockTypes.includes( 'core/block' );
+	if ( blockTypeIsDisabled ) {
+		return null;
+	}
+
+	const referencedBlockType = getBlockType( reusableBlock.type );
+	if ( ! referencedBlockType ) {
+		return null;
+	}
+
+	return {
+		name: 'core/block',
+		initialAttributes: { ref: reusableBlock.id },
+		title: reusableBlock.title,
+		icon: referencedBlockType.icon,
+		category: 'reusable-blocks',
+		keywords: [],
+		isDisabled: false,
+	};
+}
+
+/**
+ * Determines the items that appear in the the inserter. Includes both static
+ * items (e.g. a regular block type) and dynamic items (e.g. a reusable block).
+ * 
+ * @param {Object}                  state             Global application state.
+ * @param {string[]|boolean}        enabledBlockTypes Enabled block types, or true/false to enable/disable all types.
+ * @returns {Editor.InserterItem[]}                   Items that appear in inserter.
+ */
+export function getInserterItems( state, enabledBlockTypes = true ) {
+	if ( ! enabledBlockTypes ) {
+		return [];
+	}
+
+	const staticItems = getBlockTypes().map( blockType =>
+		buildInserterItemFromBlockType( state, enabledBlockTypes, blockType )
+	);
+
+	const dynamicItems = getReusableBlocks( state ).map( reusableBlock =>
+		buildInserterItemFromReusableBlock( enabledBlockTypes, reusableBlock )
+	);
+
+	const items = [ ...staticItems, ...dynamicItems ];
+	return compact( items );
```

Comment:
Can we add a unique `id` to each item. It could equal the `name` for regular blocks and the `ref` for reusable blocks. Having a unique id across all items would simplify logic in the inserter.

Reply:
👍 fixed in ef3bbc97f96bc9c5516d7501c4039fc7a70dc47b.